fabtests/unit/mr_cache_evict.c

```diff
@@ -0,0 +1,695 @@
+	int ret;
+
+	/* If a reuse address is defined, request MAP_FIXED to require the mmap
+	 * allocation to reuse this address.
+	 */
+	if (reuse_addr)
+		flags |= MAP_FIXED;
+
+	mr = mmap(reuse_addr, alloc_size, PROT_READ | PROT_WRITE, flags, -1, 0);
+	if (mr == MAP_FAILED) {
+		FT_UNIT_STRERR(err_buf, "mmap failed", -errno);
+		return NULL;
+	}
+
+	mr->type = MMAP;
+	mr->size = alloc_size;
+
+	FT_DEBUG("Allocated %p of size %lu", mr, mr->size);
+
+	if (geteuid() == 0) {
+		ret = virt_to_phys(mr, &phys_addr);
+		if (!ret)
+			FT_DEBUG("va %p -> phys addr 0x%lx", mr, phys_addr);
+	}
+
+	/* Cache this virtual address to reuse for future allocations. */
+	reuse_addr = mr;
+
+	return &mr->user;
+}
+
+/* Generic allocation/deallocation function. Only a single allocation of any
+ * type should be outstanding.
+ */
+static void mem_free(void *buf)
+{
+	struct memory_region *mr =
+		(struct memory_region *)((intptr_t)buf - offsetof(struct memory_region, user));
+
+	switch (mr->type)
+	{
+	case SBRK:
+		sbrk_free(mr);
+		break;
+	case MMAP:
+		mmap_free(mr);
+		break;
+	case BRK:
+		brk_free(mr);
+		break;
+	default:
+		break;
+	}
+}
+
+static void *mem_alloc(enum mem_type type, size_t size)
+{
+	switch (type)
+	{
+	case SBRK:
+		return sbrk_alloc(size);
+	case MMAP:
+		return mmap_alloc(size);
+	case BRK:
+		return brk_alloc(size);
+	default:
+		return NULL;
+	}
+}
+
+static double timespec_diff_to_double(const struct timespec *start,
+				      const struct timespec *end)
+{
+	return (end->tv_sec - start->tv_sec) +
+		(end->tv_nsec - start->tv_nsec) / 1E9;
+}
+
+/* MR allocation function which returns the MR and the time in seconds to
+ * allocate the MR.
+ */
+static int mr_alloc(const void *buf, size_t mr_buf_size, struct fid_mr **mr,
+		    double *alloc_time)
+{
+	uint64_t access = ft_info_to_mr_access(fi);
+	struct timespec start;
+	struct timespec end;
+	int ret;
+
+	ret = clock_gettime(CLOCK_MONOTONIC, &start);
+	if (ret) {
+		FT_UNIT_STRERR(err_buf, "clock_gettime failed", -errno);
+		return -errno;
+	}
+
+	ret = fi_mr_reg(domain, buf, mr_buf_size, access, 0, FT_MR_KEY, 0, mr,
+			NULL);
+	if (ret != FI_SUCCESS) {
+		FT_UNIT_STRERR(err_buf, "fi_mr_reg failed", -errno);
+		return -errno;
+	}
+
+	ret = clock_gettime(CLOCK_MONOTONIC, &end);
+	if (ret) {
+		FT_UNIT_STRERR(err_buf, "clock_gettime failed", -errno);
+		return -errno;
+	}
+
+	*alloc_time = timespec_diff_to_double(&start, &end);
+
+	return 0;
+}
+
+/* Run a test verifing the eviction MR cache entries. The following is how the
+ * test works:
+ * 1. Prime CPU caches by allocating a priming MR. This MR is not used for
+ *    cache measurements.
+ * 2. Allocate a buffer using mem_alloc() with either MMAP, BRK, or SBRK. The
+ *    mem_alloc() allocator is constructed to return the same virtual address
+ *    during reallocation.
+ * 3. Measure MR allocation time of the mem_alloc() buffer. Since this buffer
+ *    does not exist in the cache, this should be a long allocation time.
```

Comment:
The cache is a memory registration cache, not a memory allocation cache.  The comment here and below is confusing.  Please separate out the discussion of memory allocation, versus registration, and clearly indicate at which step each is occurring.